[WIP]Implemented LIL format for sparse array

kangzhiq opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
The idea comes from the standard format of sparse matrix. Would it be interesting as well to implement such data structure in sparse array so that some operations can be more efficient? The current format that Sparse Array module uses is DOK.

The format LIL is efficient in incremental construction. 
This experimental implementaiton of LIL algorithm will help verify the possibility and advantage of implementing different format to sparse array.  

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* tensor
  * Added algorithm for data strucutre of sparse array 
<!-- END RELEASE NOTES -->

Commit:
Implemented LIL format for sp array

- Added new class lil, initial version

kangzhiq commented on sympy/tensor/array/sp_utils.py:
```diff
@@ -0,0 +1,123 @@
+
+        if isinstance(iterable, SparseNDimArray):
+            new_dict = iterable._sparse_array
+        # if it is a dense array, it would be cast to a default sparse array
```
The idea comes from SciPy where all sparse array are cast to a CSR format before converting to other format.

kangzhiq commented on sympy/tensor/array/sp_utils.py:
```diff
@@ -0,0 +1,123 @@
+            raise NotImplementedError("Data type not yet supported")
+
+        # Add non zero value to internal list.
+        # This operation can be simplified once Array module supports tuple index
```
This is related to the problem of NumPy-like behavior in Array module. What I was trying to do is:
```python
>>> A = Array(iterable, (3, 4, 3))
>>> index = (1, 2, 0)
>>> A[index[:-1]]
```
which is suppposed to return A[1, 2, :]. 
But currently, Array module in SymPy doesn't handle the tuple as I wished.

Upabjojr replied:
It should be extended.

Upabjojr replied:
I mean... if `index[:-1]` is not supported... maybe support for it should be added.

kangzhiq commented on sympy/tensor/array/sp_utils.py:
```diff
@@ -0,0 +1,123 @@
+    1
+    >>> a[0, 1] = 3
+    >>> a.tolist()
+    [[0, 3, 0], [2, 0, 0]]
```
Only some basic operations are tested. I have briefly tested addition and multiplication, it didn't work but there was not an error either. I will check it later. :-)

czgdp1807 commented on sympy/tensor/array/sp_utils.py:
```diff
@@ -0,0 +1,123 @@
+    """
+    def __new__(cls, iterable=None, shape=None, **kwargs):
+        shape, flat_list = cls._handle_ndarray_creation_inputs(iterable, shape, **kwargs)
+        shape = Tuple(*map(_sympify, shape))
```
`Tuple.fromiter` possible?

kangzhiq replied:
Yes, thank you for this advice!

Upabjojr commented:
I see you have implemented the whole class. I was expecting just the algorithm.

Commit:
Update implementation of format

- Ameliorated implementation of lil
- Added implementation of coo after testing
- Added basic implementation of csr

Commit:
Basic structure of CSR format

- Implemented basic structure of csr format
- Added some illustration for each format

Upabjojr commented on sympy/tensor/array/sp_utils.py:
```diff
@@ -0,0 +1,345 @@
+    return all([shape[i] > index[i] for i in range(len(shape))])
+
+# Row-based linked list sparse matrix(LIL)
+class lil(Basic, SparseNDimArray):
```
We need a better name.

kangzhiq replied:
What about LilArray? Or maybe LilSparseArray

Upabjojr commented on sympy/tensor/array/sp_utils.py:
```diff
@@ -0,0 +1,345 @@
+            return 0
+
+    # For mutable arrays
+    def __setitem__(self, index, value):
```
This should raise an exception if the class inherits `Basic`

kangzhiq replied:
Yes, it is a little bit contradictory to handle both immutable and mutable property in one class. I was thinking that we can handle this property by adding a flag: `isMutable`, but it turns out that it is not very pratical.

Upabjojr replied:
No, it cannot be done. The reason is complicated but SymPy needs all of its objects to be immutable. If you want a mutable object, you need to create a separate class.

Have a look at the existing class structures.

Upabjojr commented:
Where are the tests?

Upabjojr commented:
If the class is mutable, it should not be a SymPy object.

I see a lot of code that could be better reorganized.

kangzhiq commented:
I am working on the amelioration of code and adding the tests

kangzhiq commented:
> Where are the tests?
> I see a lot of code that could be better reorganized.

I am working on the amelioration of code and adding the tests

Commit:
Added tests and updated codes

- Added tests for new class
- Ameliorate code for new class

Commit:
Added new constructor and convertion of format

- Added new way to initialize a new format, which can facilitate the
convertion of different format
- Added an example of converint COO to CSR
- Added tests accordingly

Upabjojr commented on sympy/tensor/array/sp_utils.py:
```diff
@@ -0,0 +1,396 @@
+        return self.toscr()._mul(other)
+
+    def __setitem__(self, index, value):
+        raise TypeError('Immutable N-dim array')
```
what about moving this to the `Immutable...` superclass?

kangzhiq replied:
Yes, this line of code is also used in `ImmutableDenseNDimArray`, `ImmutableSparseNDimArray` , removing it to `ImmutableNDimArray` will avoid repetitive code.

Commit:
Added new conversion of format

- Removed inheritance of Basic
- Added conversion from LIL to CSR format
- Added tests accordingly

Commit:
Removed args test

Upabjojr commented:
There is a conflict with the current master.

czgdp1807 commented:
Looks like a good work. Attaching `Please take over` label to it.
